src/util/psbt/mod.rs

```diff
@@ -196,6 +205,467 @@ impl PartiallySignedTransaction {
 
         Ok(())
     }
+
+    /// Attempts to create _all_ the required signatures for this PSBT using `k`.
+    ///
+    /// If you just want to sign an input with one specific key consider using `sighash_ecdsa`. This
+    /// function does not support scripts that contain `OP_CODESEPARATOR`.
+    ///
+    /// # Panics
+    /// If an attempt is made to sign a taproot input.
+    pub fn sign<C: Signing>(&mut self, k: impl GetKey + Copy, secp: &Secp256k1<C>) -> Result<(), SignError> {
+        let tx = self.unsigned_tx.clone(); // clone because we need to mutably borrow when signing.
+        let mut cache = SighashCache::new(&tx);
+
+        for i in 0..self.inputs.len() {
+            match self.signing_algorithm(i).expect("i is valid") {
+                SigningAlgorithm::Ecdsa => self.bip32_sign_ecdsa(k, i, &mut cache, secp)?,
```

Comment:
I think we should try to sign all inputs instead of early return with `?`.

Reply:
Interesting, what sort of return would you imagine?
- The last error if there was one
- A vector of errors
- A map of index -> error for errors mapped by input_index
- An error type that models "3 of 6 inputs errored during signing"

Reply:
We should convey which inputs we did not sign for. 

I am also not sure whether we should error when the key is not found. It is likely that in a multi-party setting, we don't the secrets of the other party. When we can sign on psbt, we should sign all the inputs that we can sign for. In case, there are some inputs that we cannot sign for, we should still Ok() as the operation is still successful. 

How about the following:
1) Instead of `Err(SignError::KeyNotFound);`, we `continue` if the key is not found. I will mention this separately as a review comment.
2) Update the documentation suggesting that we only sign inputs that we can sign for. 

3a) Change the return type to `Result<(), BtreeMap<index, Error>>`. Also, suggest that the user can iterate over the partial_sigs to check which keys were signed by the API. This matches how the signing is done in the bitcoin core psbt wallet. 
OR 
3b) Make the return type `Result<BtreeMap<index, Vec<Pk>>, (BTreeMap<index, Vec<Pk>>, BTreeMap<index, Error>)>`. This IMO would be an ideal API, but it looks complicated. The `BtreeMap<index, Vec<Pk>>` represents a map from the input index to a list of public keys that were signed. And `BTreeMap<index, Error>)` would represent the map from the index to the error that occurred.

I like 3b) I feel that this conveys all the information about the success/failure of the API that the user should care about. It also does not return an error in case we cannot sign for input because we do not know the secret key. The information about which keys were signed is correctly conveyed to the users. The user can correctly handle the case if a key they expected to sign is not signed.

Reply:
@tcharding, this was not addressed in the latest force push. Let me know if this needs more elaboration.

Reply:
It also would be handy to break this into two public APIs. `pbst_sign_input` which additionally takes in an index to sign. That way, advanced users can have more fine-grained control of which inputs to sign.

Reply:
I've implemented the errors as suggested, thanks. 

> It also would be handy to break this into two public APIs. pbst_sign_input which additionally takes in an index to sign. That way, advanced users can have more fine-grained control of which inputs to sign.

I think `sighash_ecdsa()` is enough for users that want to sign a particular input since signing is a one line function call. The docs point folks to that method.

Reply:
The signing call in the example is pretty funky now, posting here for folks to see easily:
```rust
    /// Signs `psbt` with this signer.
    fn sign_psbt<C: Signing>(&self, secp: &Secp256k1<C>, mut psbt: Psbt) -> Result<Psbt> {
        match psbt.sign(self.master_xpriv, secp) {
            Ok(keys) => assert_eq!(keys.len(), 1),
            Err((_, e)) => {
                let e = e.get(&0).expect("at least one error");
                return Err(Error::PsbtSign(e.clone()));
            }
        };
        Ok(psbt)
    }
```